ccl/changefeedccl: TestChangefeedSchemaTTL failed

**Describe the problem**

Please describe the issue you observed, and any steps we can take to reproduce it:

```
Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:6534
            	            				pkg/ccl/changefeedccl/helpers_test.go:1524
            	Error:      	Expect "ERROR: internal error: cannot advance frontier to 1750959678.165615000,0, which is earlier than the current frontier 1750959678.165804000,0 (SQLSTATE XX000)" to match "GC threshold"
```

**To Reproduce**

checkout 65468cffb3993e61b3bcda2fb966b1126e4a9bd0

Force the test feed sink to be enterprise for 
`TestChangefeedSchemaTTL`

What did you do? Describe in your own words.

```
./dev test pkg/ccl/changefeedccl --filter TestChangefeedSchemaTTL --stress --deadlock
```

Jira issue: CRDB-51896

```    
test_log_scope.go:76: use -show-logs to present logs inline
    --- FAIL: TestChangefeedSchemaTTL/sinkless (35.33s)
        helpers_test.go:1241: making server as system tenant
        helpers_test.go:1317: making sinkless feed factory
        helpers_test.go:1382: pgURL sinkless SinklessFeedUser
        helpers_test.go:1382: pgURL sinkless root
        helpers_test.go:1053: did not force enriched envelope for CREATE CHANGEFEED FOR TABLE foo because *changefeedccl.sinklessFeedFactory is not supported
        testfeed_test.go:294: sinkless feed creating changefeed: CREATE CHANGEFEED FOR TABLE foo
        testfeed_test.go:245: sinkless feed closing connection due to timeout (30s): CREATE CHANGEFEED FOR TABLE foo
        changefeed_test.go:6534: 
            	Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:6534
            	            				pkg/ccl/changefeedccl/helpers_test.go:1524
            	Error:      	Expect "read tcp 127.0.0.1:63000->127.0.0.1:62959: use of closed network connection" to match "GC threshold"
            	Test:       	TestChangefeedSchemaTTL/sinkless
        testfeed_test.go:301: closing sinkless feed
    changefeed_test.go:6544: -- test log scope end --
```

```
--- FAIL: TestChangefeedSchemaTTL (5.13s)
    test_log_scope.go:165: test logs captured to: /Users/keithchow/go/src/github.com/cockroachdb/cockroach/y/_tmp/f71b2c569099dbed3eb71606634940c1/logTestChangefeedSchemaTTL1473415331
    test_log_scope.go:76: use -show-logs to present logs inline
    --- FAIL: TestChangefeedSchemaTTL/webhook (5.07s)
        helpers_test.go:1241: making server as system tenant
        helpers_test.go:1317: making webhook feed factory
        helpers_test.go:1103: forcing enriched envelope for *changefeedccl.webhookFeedFactory - EXPERIMENTAL CHANGEFEED FOR TABLE foo WITH OPTIONS (envelope = 'enriched', enriched_properties = 'source')
        changefeed_test.go:6534: 
            	Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:6534
            	            				pkg/ccl/changefeedccl/helpers_test.go:1524
            	Error:      	Expect "cannot advance frontier to 1750960936.535308000,0, which is earlier than the current frontier 1750960936.541562000,0" to match "GC threshold"
            	Test:       	TestChangefeedSchemaTTL/webhook
    changefeed_test.go:6544: -- test log scope end --
```

Failed on GCE Worker
```
--- FAIL: TestChangefeedSchemaTTL (26.68s)
    test_log_scope.go:165: test logs captured to: /home/keith_chow_cockroachlabs_com/go/src/github.com/cockroachdb/cockroach/y/_tmp/b3c5fea154e7bdc26bac241091cd0e95/logTestChangefeedSchemaTTL3676962245
    test_log_scope.go:76: use -show-logs to present logs inline
    --- FAIL: TestChangefeedSchemaTTL/enterprise (26.66s)
        helpers_test.go:1241: making server as system tenant
        helpers_test.go:1317: making enterprise feed factory
        helpers_test.go:1331: pgURL enterprise root
        helpers_test.go:1053: did not force enriched envelope for CREATE CHANGEFEED FOR TABLE foo because *changefeedccl.tableFeedFactory is not supported
        changefeed_test.go:6534: 
            	Error Trace:	pkg/ccl/changefeedccl/changefeed_test.go:6534
            	            				pkg/ccl/changefeedccl/helpers_test.go:1524
            	Error:      	Expect "cannot advance frontier to 1750989632.798960752,0, which is earlier than the current frontier 1750989632.801320877,0" to match "GC threshold"
            	Test:       	TestChangefeedSchemaTTL/enterprise
    changefeed_test.go:6544: -- test log scope end --
FAIL
I250627 02:00:34.152229 1 (gostd) testmain.go:685  [-] 1  Test //pkg/ccl/changefeedccl:changefeedccl_test exited with error code 1
```

--- a/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
+++ b/pkg/ccl/changefeedccl/schemafeed/schema_feed.go
@@ -633,6 +633,16 @@ func (tf *schemaFeed) ingestDescriptors(
 	return tf.adjustTimestamps(startTS, endTS, validateErr)
 }
 
+// frontierAdvanceCheckEnabled controls whether the changefeed will
+// attempt to advance the frontier depending on the relation between the
+// last recorded frontier and the current time.
+var frontierAdvanceCheckEnabled = settings.RegisterBoolSetting(
+	settings.ApplicationLevel,
+	"changefeed.frontier_advance_check.enabled",
+	"if true, attempts to advance the frontier only if the last recorded frontier is less than the current time",
+	true,
+)
+
 // adjustTimestamps adjusts the frontier or error timestamp appropriately.
 func (tf *schemaFeed) adjustTimestamps(startTS, endTS hlc.Timestamp, validateErr error) error {
 	tf.mu.Lock()
@@ -644,9 +654,15 @@ func (tf *schemaFeed) adjustTimestamps(startTS, endTS hlc.Timestamp, validateErr
 		}
 		return validateErr
 	}
-
-	if frontier := tf.mu.ts.frontier; frontier.Less(startTS) {
+	frontier := tf.mu.ts.frontier
+	if frontier.Less(startTS) {
 		return errors.Errorf(`gap between %s and %s`, frontier, startTS)
+	}
+	// If the current frontier is greater than the endTS,
+	// then we do not need to advance the frontier. In fact,
+	// trying to advance the frontier could result in an error.
+	if endTS.LessEq(frontier) && frontierAdvanceCheckEnabled.Get(&tf.settings.SV) {
+		return nil
 	}
 	return tf.mu.ts.advanceFrontier(endTS)
 }

